cupy/sorting/search.py

```diff
@@ -181,7 +182,155 @@ def where(condition, x=None, y=None):
     return _where_ufunc(condition.astype('?'), x, y)
 
 
-# TODO(okuta): Implement searchsorted
+# This is to allow using the same kernels for all dtypes, ints & floats
+# as nan is a special case
+_preamble = '''
+template<typename T>
+__device__ bool _isnan(T val) {
+    return false;
+}
+template<>
+__device__ bool _isnan<float16>(float16 val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<float>(float val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<double>(double val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<const complex<double>&>(const complex<double>& val) {
+    return isnan(val);
+}
+template<>
+__device__ bool _isnan<const complex<float>&>(const complex<float>& val) {
+    return isnan(val);
+}
+'''
+
+_searchsorted_kernel_left = core.ElementwiseKernel(
+    'S x, raw T bins, int32 n_bins',
+    'U y',
+    '''
+    y = 0;
+    if (_isnan<S>(x) || x > bins[n_bins-1]) {
+        y = n_bins;
+        return;
+    }
+    size_t l = 0;
```

Comment:
I found `l` really difficult to distinguish against `1`.
Let's use another name.
I think `left` and `right` are good.